### PR TITLE
EstimateTradeBasedOnPriceImpact Query

### PR DESCRIPTION
**Pull request by VitalyVolozhinov**
## What is the purpose of the change

> Add a description of the overall background and high level changes that this PR introduces

Added an `EstimateTradeBasedOnPriceImpact` query to the `x/poolmanager/query_proto_wrap.go`, the goal of this query is to help external blockchains/smart-contracts to fill in the `MsgSwapExactAmountIn` details by retrieving input and output values for a trade. Given a total input amount, a maximum price impact, and an external price reference this query should return an input and output trade that respects the price impact of a pool.

This query operates in a binary search way where inputs are recursively tried and reduced each time if the price impact of the output is not met. If no possible trade is found then the outputs would be zero.

## Testing and Verifying

This change added tests and can be verified as follows:

  - *Added unit test that validates the query reacts as needed for the Balancer/StableSwap and Concentrated liquidity pool types under different input conditions.

## Documentation and Release Note

  - [X] Does this pull request introduce a new feature or user-facing behavior changes?
  - [x] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [X] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [X] Code comments?
  - [ ] N/A

### CODE REVIEW

**Comment by VitalyVolozhinov** [comment]
> @VitalyVolozhinov Thanks for the PR! Would you be able to rebase your branch to current main so we have a clean git history for this PR?

The rebase is messy causing some breaking code changes for me.

**Comment by mattverse** [comment]
@VitalyVolozhinov No worries, I can review it as of rn!

File: x/poolmanager/client/query_proto_wrap.go [hunk 1]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+
+	spotPrice, err := swapModule.CalculateSpotPrice(ctx, req.PoolId, req.FromCoin.Denom, req.ToCoinDenom)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// If ExternalPrice is specified we need to adjust the maxPriceImpact based on the deviation between spot and
+	// external price.
+	adjustedMaxPriceImpact := req.MaxPriceImpact
+	if !req.ExternalPrice.IsZero() {
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:298]: wdyt of renaming ExternalPrice to TargetSpotPrice or so?

File: x/poolmanager/client/query_proto_wrap.go [hunk 2]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+		}, nil
+	}
+
+	// Validate if the trade as is respects the price impact, if it does re-estimate it with a swap fee and return
+	// the result.
+	currTradePrice := sdk.NewDec(req.FromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+	priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:358]: Whats the reason we take abs value here? can any of these value be negative in the first place?

File: x/poolmanager/client/query_proto_wrap.go [hunk 3]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+
+		if tokenOut.IsZero() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+
+		currTradePrice := sdk.NewDec(currFromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+		priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:401]: I think we can make a separate method for this since this is re-occurring & for code readability

File: x/poolmanager/client/query_proto_wrap.go [hunk 4]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+	}
+
+	// Define low and high amount to search between. Start from 1 and req.FromCoin.Amount as initial range.
+	lowAmount := sdk.OneInt()
+	highAmount := req.FromCoin.Amount
+	currFromCoin := req.FromCoin
+
+	for lowAmount.LTE(highAmount) {
+		// Calculate currFromCoin as the new middle amount to try trade.
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:381]: can we add in line comment here explaining algorithm?

File: x/poolmanager/client/query_proto_wrap.go [hunk 5]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+		return err
+	})
+
+	// Find out if the error is because the amount is too large or too little. The calculation should error
+	// if the amount is too small, and it should panic if the amount is too large. If the amount is too large
+	// we want to continue to iterate to find attempt to find a smaller value.
+	if err != nil && !strings.Contains(err.Error(), "panic") {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:456]: What's the reason we catch panic here, not in Balancer?

File: x/poolmanager/client/query_proto_wrap.go [hunk 6]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
+		currFromCoin = sdk.NewCoin(req.FromCoin.Denom, midAmount)
+
+		err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			tokenOut, err = swapModule.CalcOutAmtGivenIn(ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec())
+			return err
+		})
+
+		// If it returns an error without a panic it means the input has become too small and we should return.
+		if err != nil && !strings.Contains(err.Error(), "panic") {
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:495]: same question here for the panic

File: x/poolmanager/client/query_proto_wrap.go [hunk 7]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+			// and we should continue halving.
+			highAmount = midAmount.Sub(sdk.OneInt())
+		} else {
+			currTradePrice := sdk.NewDec(currFromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+			priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
+
+			if priceDeviation.LTE(adjustedMaxPriceImpact) {
+				lowAmount = midAmount.Add(sdk.OneInt())
+			} else {
+				highAmount = midAmount.Sub(sdk.OneInt())
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:511]: I think alot of code here has duplicate logic with the balancer one and perhaps can be generalized into a method, can we refactor it so that they can share single method?

File: x/poolmanager/client/query_proto_wrap.go [hunk 8]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+
+	for lowAmount.LTE(highAmount) {
+		// Calculate currFromCoin as the new middle amount to try trade.
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
+		currFromCoin = sdk.NewCoin(req.FromCoin.Denom, midAmount)
+
+		tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec())
+		if err == nil {
+			// If the tokenOut was returned to be zero it means the amount being traded is too small. We ignore the
+			// error output here as it could mean that the input is too large.
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:594]: ditto here for method generalization and removing code duplication

File: x/poolmanager/router_test.go [hunk 9]
```diff
@@ -1544,6 +1546,482 @@ func (s *KeeperTestSuite) TestSingleSwapExactAmountIn() {
...
+			poolId:            poolId,
+			req: queryproto.EstimateTradeBasedOnPriceImpactRequest{
+				FromCoin:       sdk.NewCoin(assetBaz, sdk.NewInt(1_000_000)),
+				ToCoinDenom:    assetBar,
+				PoolId:         poolId,
+				MaxPriceImpact: maxPriceImpact,
+				ExternalPrice:  externalPriceOneBalancer,
+			},
+			expectedInputCoin:  sdk.NewCoin(assetBaz, sdk.NewInt(39_947)),
+			expectedOutputCoin: sdk.NewCoin(assetBar, sdk.NewInt(59_327)),
```
Comment (mattverse) [inline x/poolmanager/router_test.go:1616]: Can we have one line comment on how we're getting these numbers if possible?

**Review by mattverse — COMMENTED** [review]
Core logic lgtm! 
I was actually thinking of creating this same feature myself, I think this feature is what we have really needed for ages, and would improve QOL significantly! Super excited about this. 

Left some minor comments, please take a look:)

File: proto/osmosis/poolmanager/v1beta1/query.proto [hunk 10]
```diff
@@ -248,4 +256,24 @@ message TotalLiquidityResponse {
...
   ];
-}
\ No newline at end of file
+}
+
+//=============================== EstimateTradeBasedOnPriceImpact
+message EstimateTradeBasedOnPriceImpactRequest {
+  cosmos.base.v1beta1.Coin from_coin = 1 [ (gogoproto.nullable) = false ];
+  string to_coin_denom = 2;
+  uint64 pool_id = 3;
```
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply proto/osmosis/poolmanager/v1beta1/query.proto:265]: Added

File: x/poolmanager/README.md [hunk 11]
```diff
@@ -330,3 +330,130 @@ higher slippage.
 
 Note, that the actual split happens off-chain. The router is only responsible for executing the swaps in the order and quantities of token in provided
 by the routes.
+
```
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply x/poolmanager/README.md:333]: :fire: :fire:

File: x/poolmanager/README.md [hunk 12]
```diff
@@ -330,3 +330,130 @@ higher slippage.
...
+6. Then according to the pool type we attempt to find a viable trade, we must process each pool type differently as they return different results for different scenarios. The sections below explain the different pool types and how they each handle input.
+
+
+#### Balancer Pool Type Process
+
+The following is the example input/output when executing `CalcOutAmtGivenIn` on balancer pools:
+
+- If the input is greater than the total liquidity of the pool, the output will be the total liquidity of the target token.
+- If the input is a normal amount, the output will be a normal amount, based on pool data.
+- If the input is a small amount e.g `1`, the output will be `1`, regardless of slippage.
```
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply x/poolmanager/README.md:371]: Adjusted comments:
```
- If the input is an amount that is reasonably within the range of liquidity of the pool, the output will be a tolerable slippage amount based on pool data.
- If the input is a small amount for which the pool cannot calculate a viable swap output e.g `1`, the output will be `1`, regardless of slippage.
```

File: x/poolmanager/client/query_proto_wrap.go [hunk 13]
```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+	}
+
+	if req.FromCoin.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid from coin denom")
+	}
+
+	if req.ToCoinDenom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid to coin denom")
+	}
+
```
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:279]: Since I'm calling `swapModule.CalculateSpotPrice` directly instead of using the router.go func this logic would need to be moved down to each individual `SpotPrice` calculation, in other queries this check is done before that:
```
// SpotPrice returns target pool asset prices on base and quote assets.
// nolint: staticcheck
func (q Querier) SpotPrice(ctx context.Context, req *types.QuerySpotPriceRequest) (*types.QuerySpotPriceResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "empty request")
	}

	if req.BaseAssetDenom == "" {
		return nil, status.Error(codes.InvalidArgument, "invalid base asset denom")
	}

	if req.QuoteAssetDenom == "" {
		return nil, status.Error(codes.InvalidArgument, "invalid quote asset denom")
	}

	sdkCtx := sdk.UnwrapSDKContext(ctx)

	// Note: the base and quote asset argument order is intentionally incorrect
	// due to a historic bug in the original implementation.
	sp, err := q.Keeper.CalculateSpotPrice(sdkCtx, req.PoolId, req.BaseAssetDenom, req.QuoteAssetDenom)
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}

	return &types.QuerySpotPriceResponse{
		SpotPrice: sp.String(),
	}, nil
}
```
I would suggest leaving as is as I don't know what else relies on the `CalculateSpotPrice` func

File: x/poolmanager/client/query_proto_wrap.go (hunk 1 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:298]: I was thinking that, but in our use case we don't even use a spot price we use a TWAP price, I prefer leaving it as `ExternalPrice` as it could be anything e.g Oracle/TWAP etc

File: x/poolmanager/client/query_proto_wrap.go (hunk 2 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:358]: I haven't found a value that can be negative, I can remove it but I am not sure if there is a pool configuration that could give more tokens than inputted especially in the balancer pool type and how its weights work.

File: x/poolmanager/client/query_proto_wrap.go (hunk 5 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:456]: Balancer Pool does not panic, while StableSwap does because of the maths involved.

File: x/poolmanager/client/query_proto_wrap.go (hunk 3 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:401]: Added

File: x/poolmanager/client/query_proto_wrap.go (hunk 4 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:381]: added

File: x/poolmanager/client/query_proto_wrap.go (hunk 6 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:495]: Balancer Pool does not panic, while StableSwap does because of the maths involved.

File: x/poolmanager/client/query_proto_wrap.go (hunk 7 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:511]: I've started doing it in one algorithm but it becomes very hard to follow because each pool reacts differently with errors and panics. I can change it but I suggest against it.

File: x/poolmanager/client/query_proto_wrap.go (hunk 8 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:594]: I've started doing it in one algorithm but it becomes very hard to follow because each pool reacts differently with errors and panics. I can change it but I suggest against it.

File: cmd/osmosisd/cmd/stargate-query.go [hunk 14]
```diff
@@ -406,6 +406,8 @@ func GetStructAndFill(queryPath, module, structName string, structArguments ...s
 			v.BaseAssetDenom = structArguments[1]
 			v.QuoteAssetDenom = structArguments[2]
 			return v, nil
+		case "EstimateTradeBasedOnPriceImpactRequest":
+			return nil, fmt.Errorf("swap route parsing not supported yet")
```
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply cmd/osmosisd/cmd/stargate-query.go:410]: I m not sure what this means, shall I remove this?

**Comment by VitalyVolozhinov** [comment]
@mattverse pinging as a reminder, for re-review thank you!

File: cmd/osmosisd/cmd/stargate-query.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply cmd/osmosisd/cmd/stargate-query.go:411]: no, was a note to myself and other reviewers!

File: x/poolmanager/client/query_proto_wrap.go (hunk 2 above)
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go]: Ah I see your concern here, let's put a comment here explaining the Abs we take here is a saftey measure

File: x/poolmanager/client/query_proto_wrap.go (hunk 5 above)
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go:466]: Oh wow did not know this, can you add this in the comment please?

File: x/poolmanager/client/query_proto_wrap.go (hunk 6 above)
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go:518]: Same comment here for the comments!

File: x/poolmanager/client/query_proto_wrap.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go:279]: Oh I was referring to grouping all the logic in this method and moving it to router.go. I felt having a heavy logic in query_proto_wrap a little out of place. Dependency on CalculateSpotPrice should not matter!

File: x/poolmanager/client/query_proto_wrap.go [hunk 15]
```diff
@@ -254,3 +258,417 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+	if !req.ExternalPrice.IsZero() {
+		priceDeviation := spotPrice.Sub(req.ExternalPrice).Quo(req.ExternalPrice)
+		adjustedMaxPriceImpact = adjustedMaxPriceImpact.Sub(priceDeviation)
+
+		// If the adjusted max price impact is negative or zero it means the difference between spot and external
+		// already exceeds the max price impact.
+		if adjustedMaxPriceImpact.IsZero() || adjustedMaxPriceImpact.IsNegative() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:307]: Should we return an error for this?

File: x/poolmanager/client/query_proto_wrap.go [hunk 16]
```diff
@@ -254,3 +258,417 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
...
+	// anyway.
+	tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if tokenOut.IsZero() {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+		}, nil
```
Comment (mattverse) [inline x/poolmanager/client/query_proto_wrap.go:352]: Why are we not returning fromCoin here?

File: x/poolmanager/router_test.go [hunk 17]
```diff
@@ -1728,6 +1730,484 @@ func (s *KeeperTestSuite) TestSingleSwapExactAmountIn() {
...
+		sdk.NewCoin("usdc", sdk.NewInt(5000000000)),
+	)
+	clCoinsNotLiquid := sdk.NewCoins(
+		sdk.NewCoin("eth", sdk.NewInt(1)),
+		sdk.NewCoin("usdc", sdk.NewInt(1)),
+	)
+
+	// The below values have been tested and hard coded by using the `CalcOutAmtGivenIn` as it is quite hard to
+	// mathematically work these out.
+	tests := map[string]struct {
```
Comment (mattverse) [inline x/poolmanager/router_test.go:1768]: Can we add test cases for 
- Stable swap panic error catching
- adjustedMaxPriceImpact is zero or negative
- tokenOut from calcOutAmtGivenIn is zero
-

**Review by mattverse — COMMENTED** [review]
Looks good, leaving last review iteration's comments, and we should be good to merge this

File: x/poolmanager/client/query_proto_wrap.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:279]: Done

File: x/poolmanager/client/query_proto_wrap.go (hunk 15 above)
Comment (VitalyVolozhinov) [reply x/poolmanager/client/query_proto_wrap.go:307]: The idea behind 0,0 response is that at the current time, a trade cannot be estimated, while an error would indicate something going wrong which I don't think is the case.

**Comment by mattverse** [comment]
@VitalyVolozhinov Can you take a look at the branching here? Seems like there are conflicts that must be resolved before merging 👀

File: x/poolmanager/client/query_proto_wrap.go (hunk 15 above)
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go:303]: Can we add in in proto comment + any other inline comments?

File: x/poolmanager/client/query_proto_wrap.go (hunk 16 above)
Comment (mattverse) [reply x/poolmanager/client/query_proto_wrap.go]: Makes sense

**Review by mattverse — APPROVED** [review]
Nice work! This would definitely be helpful for so much people, thanks for the contribution! 

Before we merge, could we do two following things:
1. Add proto doc + docs in other places about invalid estimation returning 0 instead of erroring
2. Have manual QA of the query being added? (cli or lcd does not matter)

Once this is done this should be good to go. Again, thanks so much for this 🔥

**Comment by VitalyVolozhinov** [comment]
> Nice work! This would definitely be helpful for so much people, thanks for the contribution!
> 
> Before we merge, could we do two following things:
> 
> 1. Add proto doc + docs in other places about invalid estimation returning 0 instead of erroring
> 2. Have manual QA of the query being added? (cli or lcd does not matter)
> 
> Once this is done this should be good to go. Again, thanks so much for this 🔥

I think the new changes may have broken the `localnet`  and I cannot run it but I have previously tested the cli on CL pools doing the following:

Running the code using local-net commands:

- `make localnet-init`
- `make localnet-start`
- `make localnet-keys`
- `make localnet-cl-create-pool`
- `make localnet-cl-create-positions`

We then query the pool with various inputs to see what the responses are using the following command:

```
osmosisd query poolmanager estimate-trade-amount-in-amount-out-based-on-price-impact 10000000000000000uosmo uusdc 1 0.001 2.49
```

image

We can see as increase the input of `uosmo` it gives us appropriate input and output amounts to trade under the given slippage and twap value.

**Comment by mattverse** [comment]
@VitalyVolozhinov Can you run `make proto-all` + solve linter issues please?

**Comment by VitalyVolozhinov** [comment]
> @VitalyVolozhinov Can you run `make proto-all` + solve linter issues please?

All issues should be solved

**Comment by migueldingli1997** [comment]
Looking forward to seeing this merged 💪

**Comment by VitalyVolozhinov** [comment]
Not sure why the proto file generation failed it had passed on our side: https://github.com/entrypoint-zone/osmosis/actions/runs/6296358703/job/17091183632

**Comment by mattverse** [comment]
@VitalyVolozhinov Did you try generating proto after merging with the latest main? Protobuf gen would often times fail when generated without having the branch merged with latest main, could be the problem 🤔 Try and lmk! If it doesn't seem to work, happy to merge for now and then re-protogen myself

**Comment by VitalyVolozhinov** [comment]
> @VitalyVolozhinov Did you try generating proto after merging with the latest main? Protobuf gen would often times fail when generated without having the branch merged with latest main, could be the problem 🤔 Try and lmk! If it doesn't seem to work, happy to merge for now and then re-protogen myself

Yes, and it works on our side: https://github.com/entrypoint-zone/osmosis/actions/runs/6348615589/job/17245570892 not sure what is happening

**Comment by jasbanza** [comment]
Hi @mattverse, this looks amazing. When will it be available for CLI?
